nemo_aligner/data/nlp/builders.py

```diff
@@ -262,38 +262,20 @@ def build_dataset(index, name):
 
 def build_sft_dataset(data_cfg, tokenizer, num_samples, answer_only_loss=True, is_chat=True, special_tokens=None):
     dataset_cls = GPTSFTChatDataset if is_chat else GPTSFTDataset
-    dataset = dataset_cls(
-        file_path=data_cfg.file_path,
-        tokenizer=tokenizer,
-        max_seq_length=data_cfg.max_seq_length,
-        min_seq_length=data_cfg.min_seq_length,
-        add_bos=data_cfg.get("add_bos", False),
-        add_eos=data_cfg.get("add_eos", True),
-        add_sep=data_cfg.get("add_sep", False),
-        sep_id=0,
-        max_num_samples=num_samples,
-        seed=data_cfg.get("seed", 1234),
-        label_key=data_cfg.get("label_key", "answer"),
-        answer_only_loss=answer_only_loss,
-        truncation_field=data_cfg.get("truncation_field", "text"),
-        pad_to_max_length=data_cfg.get("pad_to_max_length", False),
-        index_mapping_dir=data_cfg.get("index_mapping_dir", None),
-        prompt_template=data_cfg.get("prompt_template", None),
-        virtual_tokens=0,
-        tokens_to_generate=data_cfg.get(
-            "tokens_to_generate", 0
-        ),  # used at inference time to allocate tensor positions for tokens that will be generated by inf procedure.
-        memmap_workers=data_cfg.get(
-            "memmap_workers", None
-        ),  # used to set num. of workers to create the memmap index files
-        hf_dataset=data_cfg.get(
-            "hf_dataset", False
-        ),  # Whether to load the json file with the HuggingFace dataset. otherwise, will load the jsonl file with the JSONLMemMapDataset.
-        truncation_method=data_cfg.get(
-            "truncation_method", "right"
-        ),  # used to choose truncation method. Options: ['random', 'left', 'right']
-        special_tokens=special_tokens,
-    )
+    # parameters that don't come from the config
+    dataset_params = {
+        "tokenizer": tokenizer,
+        "answer_only_loss": answer_only_loss,
+        "special_tokens": special_tokens,
+        "max_num_samples": num_samples,
+        "sep_id": 0,
+        "virtual_tokens": 0,
+    }
+    dataset_params.update(data_cfg)
+    # parameters that are in the config, but not in the dataset's init
+    for param_name in ["global_batch_size", "micro_batch_size", "shuffle", "drop_last"]:
```

Comment:
It may be safer to specify which parameters are in the dataset class rather than which ones aren't. This is because we don't have much control on the input config and sometimes we inherit a bunch of stuff from the way the base model was trained, that we don't care about.

Reply:
I see, that makes sense. If we are going to load some old model, it's going to crash after this change I guess. Let me roll-back this to the old version then. 

The reason why I thought it would be better is to allow for adding new parameters seamlessly (only add to the DL class and config and you don't need to modify this "creation" part). Plus it will have to avoid costly typos, where some parameter in the config has incorrect name and is silently unused. In fact it seems that the current SFT config in aligner has a bunch of such parameters that I removed - they are there but never used in the code which is confusing for users.

Reply:
I had in mind issues with newer models actually, whose data config may contain some new keys we encounter for the first time.

I see your point but I'm afraid that currently we need to live with the fact that there are going to be some parameters in the config that we're not going to use, because the starting point is the model config which may contain pretty much anything depending on how this model has been trained. For instance when we load an SFT model into PPO there are a bunch of SFT-specific parameters that we completely ignore.

I can't think of a really good generic solution to this issue given how the config system is currently steup. I guess we could try to maintain a list of "known parameters to ignore" and warn about unexpected ones, but that would be somewhat cumbersome and I'm not sure it would be worth the trouble.

In this specific case, it would have probably been better if the parameters that are provided as input to the dataset class had been isolated in their own subconfig, so that we could trust a 1:1 match between this subconfig and the dataset class. But it's probably too late to refactor the config of our SFT models...